kernel/object_validate.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <kernel.h>
+#include <string.h>
+#include <misc/printk.h>
+#include <kernel_structs.h>
+
+
+/**
+ * Kernel object validation function
+ *
+ * Retrieve metadata for a kernel object. This function is implemented in
+ * the gperf script footer, see gen_kobject_list.py
+ *
+ * @param obj Address of kernel object to get metadata
+ * @return Kernel object's metadata, or NULL if the parameter wasn't the
+ * memory address of a kernel object
+ */
+__weak _GENERIC_SECTION(.kobject_data.text)
+struct _k_object *_k_object_find(void *obj)
+{
+	/* This is the __weak stub implementation, the version created by
+	 * gperf will replace it in build phase 2
+	 */
+	ARG_UNUSED(obj);
+	return NULL;
+}
+
+#ifdef CONFIG_PRINTK
+const char *otype_to_str(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return "k_alert";
+	case K_OBJ_DELAYED_WORK:
+		return "k_delayed_work";
+	case K_OBJ_MEM_SLAB:
+		return "k_mem_slab";
+	case K_OBJ_MSGQ:
+		return "k_msgq";
+	case K_OBJ_MUTEX:
+		return "k_mutex";
+	case K_OBJ_PIPE:
+		return "k_pipe";
+	case K_OBJ_SEM:
+		return "k_sem";
+	case K_OBJ_STACK:
+		return "k_stack";
+	case K_OBJ_THREAD:
+		return "k_thread";
+	case K_OBJ_TIMER:
+		return "k_timer";
+	case K_OBJ_WORK:
+		return "k_work";
+	case K_OBJ_WORK_Q:
+		return "k_work_q";
+	default:
+		return "?";
+	}
+}
+#endif
+
+
+static void _k_object_grant_access(struct _k_object *ko,
+				   struct k_thread *thread)
+{
+	/* STUB, implement */
+	ARG_UNUSED(ko);
+	ARG_UNUSED(thread);
+}
+
+
+/**
+ * grant a thread access to a kernek object
+ *
+ * The thread will be granted access to the object if the caller is from
+ * supervisor mode, or the caller is from user mode AND has permissions
+ * on the object already.
+ *
+ * @param object Address of kernel object
+ * @param thread Thread to grant access to the object
+ */
+void k_object_grant_access(void *object, struct k_thread *thread)
+{
+	struct _k_object *ko = _k_object_find(object);
+
+	if (!ko) {
+		printk("granting access to non-existent kernel object %p\n",
```

Comment:
Leftover debugging printk()?

Reply:
this is intentional, k_object_grant_access() is a public API and should informatively freak out if a bad object is passed in.

Reply:
Something we need is a more "regular" way to print warning/error messages so they are easier to monitor for (not the subject of this patch though).

Reply:
This should be guarded by CONFIG_PRINTK anyway (other places of the code do this)

Reply:
that's not correct. printk() expands to a no-op inline function if CONFIG_PRINTK=n.
the reason why some other calls are guarded is because they reference the otype_to_str() function. this function is itself guarded because GCC's -ffunction/data-sections doesn't work the way you would expect with strings, and the strings would end up in the binary anyway even if otype_to_str was never used.